[ENH] Add Support for Randomly Missing Values in EM Algorithm

Aniketsy opened the pull request:
fix #2445 #2386

malevolentshryne committed:
Feature(EM) : Handling partially missing values

ankurankan committed:
Merge branch 'dev' into dev

ankurankan committed:
Merge branch 'dev' into dev

Aniketsy committed:
Add Support for Randomly Missing Values in EM Algorithm

Aniketsy committed:
Add Support for Randomly Missing Values in EM Algorithm

Aniketsy committed:
Add Support for Randomly Missing Values in EM Algorithm

Aniketsy committed:
Add Support for Randomly Missing Values in EM Algorithm

ankurankan commented on pgmpy/tests/test_estimators/test_EM.py:
```diff
@@ -347,3 +348,25 @@ def tearDown(self):
         config.set_backend("numpy")
+
+
+class TestExpectationMaximization(unittest.TestCase):
```
Ideally, the test should start by simulating some data using known CPDs, then introduce some random NaN values in the data, and test whether the implementation is able to recover the original values (up to some tolerance limits). Please have a look at the other tests for example.

ankurankan commented on pgmpy/estimators/EM.py:
```diff
@@ -346,12 +364,12 @@ def get_parameters(
             estimator = BayesianEstimator.__new__(BayesianEstimator)
```
I think the logic in Step 3.1 of the algorithm will need to be modified, as it currently fixes all CPDs that do not involve latent variables. However, with randomly missing values, non-latent variables can also have missing values, and we should not be fixing the CPDs for them.

ankurankan commented on pgmpy/estimators/EM.py:
```diff
@@ -137,19 +136,27 @@ def _parallel_compute_weights(
+            missing_vars = [
+                var
+                for var in latent_card.keys()
+                if var not in data_unique.columns
```
I don't think we should check whether var is not in `data_unique.columns`. This line is going to get executed `n_samples * max_iter` times. Its better to do this check in the `__init__` method if it doesn't already exists.

ankurankan commented on pgmpy/estimators/EM.py:
```diff
@@ -113,13 +103,22 @@ def _get_log_likelihood(self, datapoint: Dict[str, Any]) -> float:
+                                    value, (int, float, np.integer, np.floating)
+                                )
+                                else value
+                            )
```
Why is this change required? Doesn't the logic in `_parallel_compute_weights` already make sure that there are no Nan values before the `_get_log_likelihood` call?

ankurankan commented on pgmpy/tests/test_estimators/test_EM.py:
```diff
@@ -192,7 +193,7 @@ def test_em_init_missing_data_handling(self):
 
         # Data shape and column removal
-        self.assertEqual(est.data.shape, (2, 3))
+        self.assertEqual(est.data.shape, (3, 3))
```
Why did this value change?

Aniketsy replied:
Previously, any row containing a `NaN` was removed, now only columns where all values are NaN are dropped. Please correct me if i’m misunderstanding anything.

ankurankan reviewed (changes requested):
@Aniketsy Thanks for the PR, I think there are some mistakes in the implementation. Please have a look at my comments. I also feel that the PR is generated using some help from an AI (I am seeing some unnecessary changes and checks that are not required; AI tools typically add those). If that is the case, I would suggest to be a bit more careful with the changes, and try to understand the algorithm a bit more deeply so that we don't end up merging something that outputs incorrect results.

Aniketsy commented:
@ankurankan Thanks for pointing it out. I made some changes using an llm, and while testing locally I ran into two failing tests. I think those changes were unnecessary. I’ll go through the algorithm again in depth and correct the implementation.

Aniketsy committed:
Add Support for Randomly Missing Values in EM Algorithm

ankurankan commented:
@Aniketsy, is this ready for review?

ankurankan commented on pgmpy/estimators/EM.py:
```diff
@@ -113,11 +106,19 @@ def _get_log_likelihood(self, datapoint: Dict[str, Any]) -> float:
         for cpd in self.model_copy.cpds:
             scope = set(cpd.scope())
+            if any(pd.isna(datapoint.get(var)) for var in scope):
+                continue
```
Don't think this is correct. There shouldn't be any missing values in this method. The missing values should have been filled in before being passed here.

ankurankan commented on pgmpy/estimators/EM.py:
```diff
@@ -137,19 +138,26 @@ def _parallel_compute_weights(
+                var
+                for var in latent_card.keys()
+                if pd.isna(data_unique.iloc[i].get(var))
+            ]
```
This is incorrect, `missing_vars` should also consider the variables that have missing values, not just the ones that are specified in `latent_card`, unless you preprocess the `latent_card` to contain those as well.

ankurankan reviewed (changes requested):
@Aniketsy I think the algorithm implementation is incorrect. It currently drops the (partial) row whenever it encounters Nan values. The EM algorithm shouldn't do that; instead it should iterate over all possible states of the missing values and compute a weighted dataset using it. 

If you are using an LLM to generate the code, please stop doing it if you don't fully understand the algorithm. It puts the burden of verifying the implementation on the reviewer and it takes a lot of time to review such PRs when the mistakes are cleverly hidden by the LLM. Please first try to understand what the algorithm is supposed to do. I would be happy to explain the algorithm in any of our community meetings.

Aniketsy commented:
@ankurankan I was planning to do in weekend, this was not ready, and i apologize I thought this was in draft, and I was removing changes which  I had done previously as i mentioned. I took the help of  llm earlier to fix two failures in which I was stuck.
sure I will ask to explain the algorithm in which part I have doubt or not understand.

Aniketsy commented:
Also i would like to apologize for creating extra work and causing any inconvenience.

Aniketsy committed:
Add Support for Randomly Missing Values in EM Algorithm

Aniketsy committed:
Add Support for Randomly Missing Values in EM Algorithm

Aniketsy commented:
@ankurankan I've reverted the unnecessary changes I introduced earlier and spent some time reviewing the implementation again. I would also like to understand the algorithm in more depth in any meeting, I’ve already joined the discord server, please let me know when it might be possible.

> I think the logic in Step 3.1 of the algorithm will need to be modified, as it currently fixes all CPDs that do not involve latent variables. However, with randomly missing values, non-latent variables can also have missing values, and we should not be fixing the CPDs for them.

I haven’t covered this part yet. I want to understand the algorithm more clearly before making changes there. I’ll explore this next, and if I’m confident about implementing it correctly, I’ll update the PR.

Aniketsy commented:
> I think the logic in Step 3.1 of the algorithm will need to be modified, as it currently fixes all CPDs that do not involve latent variables. However, with randomly missing values, non-latent variables can also have missing values, and we should not be fixing the CPDs for them.

@ankurankan for this i came up with this, please let me know if this needs improvement.
identify all variables in the data that have any missing values and collect them in `vars_with_missing`.
and then while building `fixed_cpd_vars`, we subtract both
- all variables with missing values
- all children of those variables

ankurankan commented:
@Aniketsy 

>  I would also like to understand the algorithm in more depth in any meeting, I’ve already joined the discord server, please let me know when it might be possible.

We have dev meetings on Mondays and community meetings on Thursdays. You would see them in the events (top left) on discord.

> identify all variables in the data that have any missing values and collect them in vars_with_missing.
and then while building fixed_cpd_vars, we subtract both

This would be the first step. Additionally, in the `_parallel_compute_weights` method, we expand the dataframe by adding rows corresponding to each possible value of the missing variable (line 142). Since now we want to support randomly missing values, we will have to check for each row here which values are missing and based on that expand the dataframe. I also think that checking for every row might slow down the algorithm, so it might be a good idea to precompute a boolean array that has True corresponding to each missing value, and this might give us a faster way to look up the missing variables in each row.

Aniketsy committed:
Add Support for Randomly Missing Values in EM Algorithm

Aniketsy committed:
Merge branch 'dev' into fix-2386

Aniketsy commented:
@ankurankan I’ve applied the changes based on your suggestions. could you please take a look and let me know if I’m heading in the right direction or if there’s anything that needs improvement.

ankurankan commented:
Tests are failing. Converting this to draft.

Aniketsy commented:
I'll revisit this shortly and fix the issues. Thanks!
